### PR TITLE
[SofaKernel] Regroup all compatibility files in a Sofa.Compat module

### PR DESCRIPTION
**Pull request by fredroy**
Factorize all compatibility code in the SofaFramework (GL, Helper, Defaulttype, Core) into a common module, to simplify maintainability.

For now, it is forcefully loaded for all the modules, as no code has been ported to the new architecture yet.

In the future, this module should be optional (and its inclusion in an other cmake file in a module/ directory for example)

(out-of-tree configuration tested with SofaPython3)
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/Sofa.Compat/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# Sofa.Config
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Compat/README.md]:
```suggestion
# Sofa.Compat
```

File: SofaKernel/modules/Sofa.Compat/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,91 @@
+cmake_minimum_required(VERSION 3.12)
+project(Sofa.Compat LANGUAGES CXX)
+
+set(SOFACOMPATSRC_ROOT "src/")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Compat/CMakeLists.txt]:
```suggestion
set(SOFACOMPATSRC_ROOT "src")
```

File: SofaKernel/modules/Sofa.Compat/src/sofa/initSofaCompat.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#include <sofa/config.h>
+#include <sofa/config/sharedlibrary_defines.h>
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Compat/src/sofa/initSofaCompat.cpp]:
```suggestion
```
sofa/config.h already includes sofa/config/sharedlibrary_defines.h

File: SofaKernel/modules/Sofa.Type/CMakeLists.txt [hunk 4]
```diff
@@ -61,14 +61,18 @@ set(SOURCE_FILES
...
 sofa_find_package(Sofa.Config REQUIRED)
 
 add_library(${PROJECT_NAME} SHARED ${HEADER_FILES} ${SOURCE_FILES})
 
 target_link_libraries(${PROJECT_NAME} PUBLIC Sofa.Config)
 target_link_libraries(${PROJECT_NAME} PUBLIC Boost::boost)
 
+# TODO1: add an option when the sofa code has been converted
+# TODO2: if cmake >= 3.13, move those lines to SofaFramework project (or a cmakefile in module/ dir)
+add_subdirectory(../Sofa.Compat ${CMAKE_CURRENT_BINARY_DIR}/Sofa.Compat)
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/CMakeLists.txt]: Why here? I would put this in the main CMakeLists, right after the line `add_subdirectory(${SOFA_KERNEL_SOURCE_DIR}/modules/Sofa.Config ${CMAKE_CURRENT_BINARY_DIR}/Sofa.Config)`

**Review by guparan — CHANGES_REQUESTED** [review]
Nice work, it will greatly ease the deprecation of all those compatibility headers :+1:
I spotted a few things that may need to be changed though :eyes:

File: SofaKernel/modules/Sofa.Type/CMakeLists.txt (hunk 4 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Type/CMakeLists.txt]: I try to avoid adding stuff in the main CmakeLists.txt but okay with your remark

**Pull request merged**